### PR TITLE
Introduce auxiliary curve

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: #

### Changelog

* Introduce a new concept: auxiliary curve (chinese: 辅助曲线). It has the following characteristics:

  - It's not property of somewhat objects but just a named floating-point quantity associated with the evaluating animation controller instance.

  - It supplies extra information about animation.

  - Its value can be varied by time -- it can be authored in animation clip in form of real number curve.

  - It will be evaluated, sampled, blended just like bone transforms.

  - Unlike bone transforms are written back into scene graph. It can be queried as animation result.

  - (**in recent future**) It can be used in animation graph. For example, used in transition condition.

This PR does:

- Provide the ability to author auxiliary curve tracks in clips.

- Provide interfaces to query the evaluation result of auxiliary curves.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/animation/animation-clip.ts [hunk 1]
```diff
@@ -243,6 +244,18 @@ export class AnimationClip extends Asset {
...
 
+    /**
+     * @zh 获取或设置该剪辑中存储的辅助曲线数组。
+     * @experimental
+     */
+    get auxiliaryCurveInfos_experimental () {
+        return this._auxiliaryCurveInfos as ReadonlyArray<AuxiliaryCurveInfo>;
+    }
+
+    set auxiliaryCurveInfos_experimental (value) {
```
Comment (SantyWang) [inline cocos/animation/animation-clip.ts]: Setting the array directly, I am worried that there will be problems. For example, it is possible that the user will use `clip.auxiliaryCurveInfos_experimental[2] = new AuxiliarayCurveInfo();` so I suggest it is better to encapsulate the API of the operation, for example, `clip.addAuxiliarayCurve( info); clip.removeAuxiliararyCurveAt(index);` and then getter returns `readonlyArray<AuxiliararyCurveInfo>`

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts [hunk 2]
```diff
@@ -153,6 +160,24 @@ class PoseScaleBinding extends PoseBindingBase implements PoseBinding<Vec3> {
...
+    constructor (private _handle: AuxiliaryCurveHandle) {
+
+    }
+
+    public destroy (): void {
+        this._handle.destroy();
+    }
+
+    public setValue (value: number, pose: Pose): void {
+        pose.metaValues[this._handle.index] = value;
```
Comment (SantyWang) [inline cocos/animation/marionette/animation-graph-animation-clip-binding.ts]: `metaValues` maybe needs a new name related to `AnxuliarayCuve`

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts [hunk 3]
```diff
@@ -267,6 +293,27 @@ function createRuntimeBindingAG (track: TrackBinding, bindContext: AnimationClip
     return undefined;
 }
 
+class AuxiliaryCurveEvaluation {
+    constructor (private _binding: AuxiliaryCurveBinding, private _curve: RealCurve) {
```
Comment (SantyWang) [inline cocos/animation/marionette/animation-graph-animation-clip-binding.ts:297]: default value is `undefined`?

File: cocos/animation/animation-clip.ts (hunk 1 above)
Comment (shrinktofit) [reply cocos/animation/animation-clip.ts]: I used to do that in other place (for example [animation graph layer](https://github.com/cocos/cocos-engine/blob/v3.7.3/cocos/animation/marionette/animation-graph.ts#L907-L928)). A problem is that there should be **at least 4** methods: get/add/remove/adjust order (even the order does not affect evaluation, just displaying). I'm also struggling with this "big 4" problem. How do you feel on this?

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts (hunk 3 above)
Comment (shrinktofit) [reply cocos/animation/marionette/animation-graph-animation-clip-binding.ts:297]: Did you mean L293? It's an unrelated code line.

File: cocos/animation/animation-clip.ts (hunk 1 above)
Comment (SantyWang) [reply cocos/animation/animation-clip.ts]: I think it's OK, a little more code, but it ensures stable results and no boundary cases.

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts (hunk 2 above)
Comment (shrinktofit) [reply cocos/animation/marionette/animation-graph-animation-clip-binding.ts]: Solved in https://github.com/cocos/cocos-engine/pull/14659/commits/29a9d7445661dffcd5ebdd44173d6628f25a6dfd .

File: cocos/animation/animation-clip.ts (hunk 1 above)
Comment (shrinktofit) [reply cocos/animation/animation-clip.ts]: Modified here: https://github.com/cocos/cocos-engine/pull/14659/commits/dcd64c1d0909a10c8bd4fb3e467fc68e86529261 . Please review.

Comment (shrinktofit) [reply cocos/animation/animation-clip.ts]: Adjusted here: https://github.com/cocos/cocos-engine/pull/14659/commits/ee8e8309e35724dcc601245f639a89110d17f16c . Please give it a review.

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts (hunk 3 above)
Comment (SantyWang) [reply cocos/animation/marionette/animation-graph-animation-clip-binding.ts:297]: Oh, I mean in Line 297, what happen if we call `new AuxiliarayCurveEvaluation()` and would `_binding` and `_curve` will be `undefined`?

**Review by SantyWang — APPROVED** [review]

File: cocos/animation/marionette/animation-graph-animation-clip-binding.ts (hunk 3 above)
Comment (shrinktofit) [reply cocos/animation/marionette/animation-graph-animation-clip-binding.ts:297]: It could never be

**Pull request merged**